solutions/usaco-945.mdx

```diff
@@ -2,16 +2,66 @@
 id: usaco-945
 source: USACO Gold 2019 US Open
 title: Snakes
-author: Sofia Yang
+author: Sofia Yang, Ryan Chou
 ---
 
 [Official Analysis (C++)](http://www.usaco.org/current/data/sol_snakes_gold_open19.html)
 
+
+Instead of keeping track of all of the different resizings, let's assume that $dp[i][j]$ stores the minimum sum of all net sizes for the first $i$ snakes after $j$ resizings. Then the total wasted space would be $dp[n][k] - T$, where $T$ is the sum of all of the groups of snakes. 
+
 ## Implementation
 
 **Time Complexity:** $\mathcal{O}(N^3)$
 
 <LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+int main() {
+	freopen("snakes.in", "r", stdin);
+	freopen("snakes.out", "w", stdout);
+	
+	int n, k;
+	cin >> n >> k;
+
+	vector<int> groups(n + 1);
+	for (int i = 1; i <= n; i++) {
+		cin >> groups[i];
+	}
+
+	// dp[i][j] = min sum of net sizes after catching first i snakes with j resizings
+	vector<vector<int>> dp(n + 1, vector<int>(k + 1, 0));
+
+	int largest_group = -1;
+	int space_used = 0;
+	for (int i = 1; i <= n; i++) {
+		largest_group = max(largest_group, groups[i]);
+		
+		/*
+		 * base case: with zero resizings, Bessie must start off with
+		 * a net large enough to catch the largest group of snakes
+		*/
+		dp[i][0] = largest_group * i;
+
+		for (int j = 1; j <= k; j++) {
+			dp[i][j] = INT32_MAX;
+			int cur = groups[i];
+			for (int c = i - 1; c >= 0; c--) {
+				// consider the case where we switch after collecting the first c groups
+				dp[i][j] = min(dp[i][j], dp[c][j - 1] + cur * (i - c));
+				cur = max(cur, groups[c]);
```

Comment:
```suggestion
			int netSize = groups[i];
			for (int c = i - 1; c >= 0; c--) {
				// consider the case where we switch after collecting the first c groups
				dp[i][j] = min(dp[i][j], dp[c][j - 1] + netSize * (i - c));
				netSize = max(netSize , groups[c]);
```

Reply:
ryan can you rename to net_size? think snake_case is convention for c++

Reply:
alright